font/src/ft/mod.rs

```diff
@@ -302,32 +294,55 @@ impl FreeTypeRasterizer {
         Ok(font_key)
     }
 
-    fn full_metrics(&self, key: FontKey) -> Result<FullMetrics, Error> {
-        let face = self.faces.get(&key).ok_or(Error::FontNotLoaded)?;
+    fn full_metrics(&self, face_load_props: &FaceLoadingProperties) -> Result<FullMetrics, Error> {
+        let size_metrics =
+            face_load_props.ft_face.size_metrics().ok_or(Error::MissingSizeMetrics)?;
 
-        let size_metrics = face.ft_face.size_metrics().ok_or(Error::MissingSizeMetrics)?;
-
-        let width = match face.ft_face.load_char('0' as usize, face.load_flags) {
-            Ok(_) => face.ft_face.glyph().metrics().horiAdvance / 64,
-            Err(_) => size_metrics.max_advance / 64,
-        } as f64;
+        let width =
+            match face_load_props.ft_face.load_char('0' as usize, face_load_props.load_flags) {
+                Ok(_) => face_load_props.ft_face.glyph().metrics().horiAdvance / 64,
+                Err(_) => size_metrics.max_advance / 64,
+            } as f64;
 
         Ok(FullMetrics { size_metrics, cell_width: width })
     }
 
+    fn load_ft_face(&mut self, path: PathBuf, index: isize) -> Result<Rc<FTFace>, Error> {
+        let mut ft_face = self.library.new_face(&path, index)?;
+        if ft_face.has_color() {
+            unsafe {
+                // Select the colored bitmap size to use from the array of available sizes
+                freetype_sys::FT_Select_Size(ft_face.raw_mut(), 0);
+            }
+        }
+
+        let ft_face = Rc::new(ft_face);
+        self.ft_faces.insert(path, Rc::downgrade(&ft_face));
+
+        Ok(ft_face)
+    }
+
     fn face_from_pattern(
         &mut self,
         pattern: &PatternRef,
-        font_id: FontID,
-        key: Option<FontKey>,
+        font_key: FontKey,
     ) -> Result<Option<FontKey>, Error> {
         if let (Some(path), Some(index)) = (pattern.file(0), pattern.index().next()) {
-            if let Some(key) = self.keys.get(&font_id) {
-                return Ok(Some(*key));
+            if self.faces.get(&font_key).is_some() {
+                return Ok(Some(font_key));
             }
 
             trace!("Got font path={:?}", path);
-            let mut ft_face = self.library.new_face(&path, index)?;
+
+            let ft_face = if let Some(face) = self.ft_faces.get(&path) {
+                if let Some(ft_face) = face.upgrade() {
+                    ft_face
+                } else {
+                    self.load_ft_face(path, index)?
+                }
+            } else {
+                self.load_ft_face(path, index)?
+            };
```

Comment:
Maybe we can do something like this here:

```rust
let face = match self.ft_faces.get(&path).and_then(|face| face.upgrade()).unwrap_or_else(|| self.load_ft_face(path, index)?;
```

I'm not sure about the exact syntax, but it should work. Alternatively if it looks nicer, we could also `match` inside of `unwrap_or_else`, but I think it should format nicely.

Reply:
It doesn't look nice, since you can't use `?` in a closure, you also can't use `ok_or_else`, since it applies random thing and you don't have a proper function to do something if none, but not an `unwrap`, like `map_none` could help, but we don't have one, so you should add it youself, which isn't helpful here and you can't do it un Rust anyways IIRC.

Reply:
Could still apply the `match` suggestion just fine.

Reply:
Oh, it was really hard to parse your one liner tbh.